### PR TITLE
ip module feature

### PR DESCRIPTION
**Pull request by felix-engelmann**
##### ISSUE TYPE

- New Module Pull Request
##### COMPONENT NAME

ip
##### ANSIBLE VERSION

```
ansible 2.0.1.0
```
##### SUMMARY

The `ip` module aimes to provide the same functionality as the linux `ip` tool. Especially setting explicit routes which is not possible with the `nmcli` module. It uses the `netlink` socket which is abstracted by the `pyroute2` library. The module is capable of assigning and removing addresses to and from interfaces as well as managing routes.

The module is built with IPv6-first in focus, but is completely backwards compatible by automatic detection.

With further development, other functionality of the `ip` tool can be easily integrated to manage links and roles as well as refine the current operations by e.g. adding options for sophisticated routing. 

This new feature is a bit dangerous, because one can easily cut the connection to the maschine by removing vital routes, but with care it is very powerful and allows to integrate the network management such as simple routing into ansible playbooks.

It removes the hacking with the `command` module which can not provide idempotency without files and is bad at distinguishing real changes from problems.

I hope the general interest for this module is as high as mine and I would be happy to maintain and continue development for this module.

### CODE REVIEW

**Comment by sivel** [comment]
Python3 syntax is a requirement for new modules.

See https://github.com/ansible/ansible-modules-core/pull/3698#issuecomment-220179520 for info about how to properly handle exceptions.

needs_revision

**Comment by gregdek** [comment]
Thanks @felix-engelmann for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by felix-engelmann** [comment]
Thank you for the hint about the compatibility module.
ready_for_review again

**Comment by gregdek** [comment]
Thanks @felix-engelmann for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by felix-engelmann** [comment]
This module needs_revision because by adding functionality it got too complex an has to change its internal structure. Also nice and consistent code with comments would improve legibility. 
I developed it into a huge mess, but it served as great proof of concept. 
Sorry for the preliminary PR. It will return in nice shape soon.

**Comment by felix-engelmann** [comment]
Finished the complete rewrite. It is way more understandable now and has a better structure to enable extension. It is ready_for_review again.

Sorry again for the inconvenience.

File: network/ip.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+            
+            routes.add_route(addr,dev,via)
+            module.exit_json(changed=True)
+                
+    elif mode == 'link':
+        
+        # add an IF to a bridge
+        if kind == 'port':
+            
+            if state=="present":
```
Comment (adrianmoisey) [inline network/ip.py]: Inconsistent spacing around the `==`

File: network/ip.py [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+                    module.fail_json(msg='valid device and bridge required')
+                if not link.is_bridge():
+                    module.fail_json(msg='bridge is not a bridge')
+                    
+                if dev.master == link.id:
+                    module.exit_json(changed=False)
+                else:
+                    dev.set_master(link)
+                    module.exit_json(changed=True)
+            if state=="absent":
```
Comment (adrianmoisey) [inline network/ip.py]: Inconsistent spacing around the `==`

**Comment by felix-engelmann** [comment]
@adrianmoisey is there a general lint specification? for example a pylintrc to have consistent formatting?

File: network/ip.py [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+    from pyroute2.netlink import AF_INET6, AF_INET
+    import ipaddress
+    HAS_PYROUTE=1
+except ImportError:
+    HAS_PYROUTE=0
+
+ip=None
+module=None
+
+def l_key(l,key):
```
Comment (svinota) [inline network/ip.py]: see also: `nlmsg.get_attr()`

e.g.::

```
In [5]: eth = ip.get_links(ip.link_lookup(ifname='enp0s25'))[0]

In [6]: eth.get_attr('IFLA_OPERSTATE')
Out[6]: 'DOWN'

In [7]: eth.get_attr('IFLA_ADDRESS')
Out[7]: 'a0:48:1c:dd:15:cd'

In [8]: eth.get_attr('IFLA_IFNAME')
Out[8]: 'enp0s25'
```

Comment (felix-engelmann) [inline network/ip.py]: The function should be more prominent in the documentation (http://docs.pyroute2.org/iproute.html) It is only used in one example on the top without description.

Comment (svinota) [inline network/ip.py]: My bad. Updated the docs a bit.

**Comment by kamsz** [comment]
I'm wondering if it wouldn't be better to split these into separate modules like ip_addr ip_link etc. It'd make the modules easier and one module wouldn't be responsible for handling multiple things leading to higher chance of breaking something during the update :)

**Comment by felix-engelmann** [comment]
The idea of @kamsz could be easily implemented, because the `mode` parameter is more or less a namespacing. It would follow the unix philosophy of one task per tool, but I made it this way to stay as close as possible to the original `ip` command.
Shall I split it into 3 modules?

**Comment by felix-engelmann** [comment]
Spliting the file in 3 modules would also mean a lot of duplicated code as the files should be self contained.

**Comment by felix-engelmann** [comment]
Splitting the modules I found a feature idea. For setting routes, it is not neccessary to specify a device, if the network is reachable through a unique interface. Therefore the requirement for a device parameter could be relaxed for the ip_route module.

**Pull request closed without merging**